wa/workloads/exoplayer/__init__.py

```diff
@@ -0,0 +1,185 @@
+                        self.logger.warn('Use "filename"parameter instead of '
+                                         '"format" to specify a different file.')
+        return self._selected_file
+
+    def init_resources(self, context):
+        ensure_directory_exists(self.video_directory)
+        self.video_files = defaultdict(list)
+        self.enum_video_files()
+        self._selected_file = None
+
+    def enum_video_files(self):
+        for filename in os.listdir(self.video_directory):
+            for format in self.parameters['format'].allowed_values:
+                if format in filename:
+                    self.video_files[format].append(os.path.join(self.video_directory, filename))
+
+    def validate(self):
+        if (self.format and self.filename) and (self.format != self.parameters['format'].default):
+            raise ConfigError('Ether format *or* filename must be specified; but not both.')
+
+    def initialize(self, context):
+        super(ExoPlayer, self).initialize(context)
+        # Grant app all permissions
+        grant_app_permissions(self.target, self.package)
+
+    def setup(self, context):
+        self.device_video_file = self.target.path.join(self.target.working_directory,
+                                                       os.path.basename(self.host_video_file))
+        if self.force_dependency_push or not self.target.file_exists(self.device_video_file):
+            self.logger.debug('Copying {} to device.'.format(self.host_video_file))
+            self.target.push(self.host_video_file, self.device_video_file, timeout=120)
+
+        super(ExoPlayer, self).setup(context)
+
+        # Prepare logcat monitor
+        self.monitor = self.target.get_logcat_monitor(REGEXPS.values())
+        self.monitor.start()
+
+    def run(self, context):
+        play_cmd = 'am start -a {} -d "file://{}"'.format(self.action, self.device_video_file)
+        self.logger.info(play_cmd)
+        self.target.execute(play_cmd)
+
+        self.monitor.wait_for(REGEXPS['start'])
+        self.logger.info('Playing media file')
+
+        line = self.monitor.wait_for(REGEXPS['duration'])[0]
+        media_duration_s = int(round(float(re.search(REGEXPS['duration'], line)
+                                           .group('duration'))))
+
+        self.logger.info('Media duration is {} seconds'.format(media_duration_s))
+
+        if self.play_duration and self.play_duration < media_duration_s:
+            self.logger.info('Waiting {} seconds before ending playback'
+                           .format(self.play_duration))
+            time.sleep(self.play_duration)
+        else:
+            self.logger.info('Waiting for playback completion ({} seconds)'
+                           .format(media_duration_s))
+            self.monitor.wait_for(REGEXPS['end'], timeout = media_duration_s + 30)
+
+        self.logger.info('Media file playback completed')
```

Comment:
This is unnecessary -- workload completion would already be indicated in the log. This is also misleading, since if you've hit the `if` condition above, you may have ended up sleeping for some time by this point, after media playback has completed